components/resolved_ts/src/cmd.rs

```diff
@@ -0,0 +1,288 @@
+                // frequently.
+                if req.get_cmd_type() != CmdType::Delete {
+                    debug!(
+                        "skip other command";
+                        "command" => ?req,
+                    );
+                }
+                continue;
+            }
+            let mut put = req.take_put();
+            match put.cf.as_str() {
+                "write" => match decode_write(put.get_key(), put.get_value()) {
+                    Some(write) => {
+                        // TODO: handle gc_fence here.
+                        let key = Key::from_encoded(put.take_key());
+                        let commit_ts = key.decode_ts().unwrap();
+                        let key = key.truncate_ts().unwrap();
+                        if is_one_pc {
+                            change_logs.insert(
+                                key.clone(),
+                                ChangeRow::OnePc {
+                                    commit_ts,
+                                    write,
+                                    value: pending_default.remove(&key),
+                                    key,
+                                },
+                            );
+                        } else {
+                            let commit_ts = if write.write_type == WriteType::Rollback {
+                                None
+                            } else {
+                                Some(commit_ts)
+                            };
+                            change_logs.insert(
+                                key.clone(),
+                                ChangeRow::Commit {
+                                    key,
+                                    commit_ts,
+                                    write,
+                                },
+                            );
+                        }
+                    }
+                    None => continue,
+                },
+                "lock" => match decode_lock(put.get_key(), put.get_value()) {
+                    Some(lock) => {
+                        let key = Key::from_encoded(put.take_key());
+                        let value = pending_default.remove(&key);
+                        let l = change_logs
+                            .insert(key.clone(), ChangeRow::Prewrite { key, lock, value });
+                        assert!(l.is_none());
+                    }
+                    None => continue,
+                },
+                "" | "default" => {
+                    let key = Key::from_encoded(put.take_key());
+                    let start_ts = key.decode_ts().unwrap();
+                    let key = key.truncate_ts().unwrap();
+                    match change_logs.get_mut(&key) {
+                        Some(ChangeRow::Prewrite { value, lock, .. }) => {
+                            assert!(value.is_none());
+                            assert_eq!(start_ts, lock.ts);
```

Comment:
How about doing these checks (`ChangeRow::Prewrite` & `ChangeRow::OnePc`) after all requests are processed, in case L115 or L140 return a `None` and not incoming put on `default`.

Reply:
If a prewrite has a short value, it will never get into this path.

Reply:
What I mean is to check whether the `CmdBatch` is completed (panic when both `ChangeRow::Prewrite::value` and `ChangeRow::Prewrite::lock::short_value` is `None` or `pending_default` is not empty after all `Request` are processed), I think this is what the `assert` here checking about.